File: CPlusPlus/RPNParser/Stack.cpp
#ifndef Stack_CPP
#define Stack_CPP
#include "Node.cpp"
#include <iostream>
using namespace std;

class Stack
{
    public:
        Stack()
        {
        }
        void push(double newValue)
        {
            Node n (newValue);
            n.setNext(top);
            top = &n;
            cout << "Data: " << (*top).getValue() << "  Address: " << top << endl;
        }
        
        double pop()
        {
            double v = (*top).getValue();
            Node nextNode = (*top).getNext();
            top = &nextNode; 
            cout << "Popped! top: " << top << endl;
            return v;
        }
    private:
        Node *top;
};

#endif

File: Java/RPNParser/Stack.java
import java.io.*;

public class Stack
{
    Node top = null;
    public Stack()
    {
    }

    public void push(double val)
    {
        Node temp = new Node(val);
        temp.setNext(top);
        top=temp;
    }

    public double pop()
    {
        double v = top.getData();
        top = top.getNext();
        return v;
    }
    
    //region Stack bonus methods by Thomas
    public boolean isEmpty()
    {
        try{
            push(pop());
            return false;
        }
        catch(Exception e){}
        return true;
    }
    public void cascadingPrint()
    {
        try{
            while(true)
                System.out.println(pop());
        }
        catch (Exception e){};
    }
    private Node getTop()
    {
        return top;
    }
    public void invert()
    {
        Stack temp = new Stack();
        try {
            while(true)
                temp.push(pop());
        } catch (Exception e) {
            top=temp.getTop();
        }
    }
    //endregion
}